File: hygmap-api/app/schemas/star.py
"""
Pydantic schemas for star data from the athyg table
"""
from pydantic import BaseModel, Field, computed_field
from typing import Optional


class StarBase(BaseModel):
    """Base star data returned in list queries"""
    id: int
    proper: Optional[str] = None
    bayer: Optional[str] = None
    flam: Optional[str] = None
    con: Optional[str] = None
    spect: Optional[str] = None
    absmag: Optional[float] = None
    mag: Optional[float] = None
    dist: Optional[float] = None
    # Optional because a star with no usable parallax has no 3D position. The import
    # clears dist/absmag/x/y/z for those (see the sentinel note in
    # db/sql/03_import_data.sql) rather than placing them at a fabricated 100 kpc.
    # List and search results exclude them, since they cannot be drawn; a direct
    # /stars/{id} request still returns them, with nulls, rather than failing.
    x: Optional[float] = None
    y: Optional[float] = None
    z: Optional[float] = None
    # Catalog IDs for display_name fallback
    hip: Optional[str] = None
    hd: Optional[str] = None
    hr: Optional[str] = None
    gj: Optional[str] = None
    cns5: Optional[str] = None
    gaia: Optional[str] = None
    tyc: Optional[str] = None
    # Fictional name (populated when world_id is provided)
    name: Optional[str] = None

    @computed_field
    @property
    def display_name(self) -> str:
        """
        The star's display name.

        This is the ONE implementation of this rule. It previously existed four times
        (here, on StarDetail, in PHP's StarFormatter, and in React's getStarDisplayName)
        and the copies disagreed: a star with both a GJ and a HIP number was called
        "GJ 1" by PHP and "HIP 439" here. See tests/fixtures/display-names.json, which
        every suite asserts against.

        Canonical order, decided 2026-07-29:

            fictional name (only when a world is selected)
            proper
            bayer + con      (both required, trimmed)
            flam + con       (both required, trimmed)
            GJ               Gliese-Jahreiss leads because this is a map of the solar
            HD               neighbourhood, so the nearby-star catalog is the most
            HIP              useful identifier to show. This used to lead with HIP.
            HR
            CNS5
            TYC
            Gaia
            spect            informative for an anonymous star
            ID <id>          last resort, so the name is never empty
        """
        # The fictional name is only populated when the query passed a world_id, so its
        # presence is the signal. Without this branch the API could not express the name
        # PHP needed, which is why PHP reimplemented the whole chain.
        if self.name:
            return self.name
        if self.proper:
            return self.proper
        # Source data carries padding on bayer/flam; trim so the three stacks agree.
        if self.bayer and self.con:
            return f"{self.bayer.strip()} {self.con.strip()}"
        if self.flam and self.con:
            return f"{self.flam.strip()} {self.con.strip()}"
        if self.gj:
            return f"GJ {self.gj}"
        if self.hd:
            return f"HD {self.hd}"
        if self.hip:
            return f"HIP {self.hip}"
        if self.hr:
            return f"HR {self.hr}"
        if self.cns5:
            return f"CNS5 {self.cns5}"
        if self.tyc:
            return f"TYC {self.tyc}"
        if self.gaia:
            return f"Gaia {self.gaia}"
        if self.spect:
            return self.spect
        return f"ID {self.id}"


class StarDetail(StarBase):
    """Detailed star data with all fields"""
    hyg: Optional[int] = None
    hip: Optional[str] = None
    hd: Optional[str] = None
    hr: Optional[str] = None
    gj: Optional[str] = None
    cns5: Optional[str] = None
    tyc: Optional[str] = None
    gaia: Optional[str] = None
    ra: Optional[float] = None
    dec: Optional[float] = None
    dist: Optional[float] = None
    mag: Optional[float] = None

    # display_name is inherited from StarBase deliberately. It used to be a second
    # hand-written copy of the chain, and the copies had already drifted -- StarBase
    # checked hr and this one did not, so an HR-only star was "HR 2491" in a list row
    # and "ID 7" in a detail row.


class StarListResponse(BaseModel):
    """Response for star list queries"""
    result: str = "success"
    data: list[StarBase]
    length: int


class StarDetailResponse(BaseModel):
    """Response for individual star queries"""
    result: str = "success"
    data: Optional[StarDetail] = None


class ProperName(BaseModel):
    """Star with proper name for dropdown"""
    id: int
    proper: str


class ProperNamesResponse(BaseModel):
    """Response for proper names list"""
    result: str = "success"
    data: list[ProperName]
    length: int


class FictionalName(BaseModel):
    """Fictional star name for dropdown"""
    star_id: int
    name: str


class FictionalNamesResponse(BaseModel):
    """Response for fictional names list"""
    result: str = "success"
    data: list[FictionalName]
    length: int


class World(BaseModel):
    """Fictional world/universe"""
    id: int
    name: str


class WorldsResponse(BaseModel):
    """Response for worlds list"""
    result: str = "success"
    data: list[World]
    length: int

File: hygmap-frontend/src/types/star.ts
/**
 * Type definitions for star data from the athyg table
 */

export interface Star {
  id: number
  proper?: string | null
  bayer?: string | null
  flam?: string | null
  con?: string | null
  spect?: string | null
  absmag?: number | null
  x: number
  y: number
  z: number
  display_name: string
  /**
   * Fictional name, populated by the API only when a world_id was requested.
   * The React UI has no world selector yet, so this is normally absent — but the
   * display-name rule accounts for it so all three stacks share one chain.
   */
  name?: string | null
}

export interface StarDetail extends Star {
  hyg?: number | null
  hip?: string | null
  hd?: string | null
  hr?: string | null
  gj?: string | null
  tyc?: string | null
  cns5?: string | null
  gaia?: string | null
  ra?: number | null
  dec?: number | null
  dist?: number | null
  mag?: number | null
}

export interface StarListResponse {
  result: string
  data: Star[]
  length: number
}

export interface StarDetailResponse {
  result: string
  data: StarDetail | null
}

export interface BoundingBox {
  xmin: number
  xmax: number
  ymin: number
  ymax: number
  zmin: number
  zmax: number
}

/**
 * Get the display name for a star.
 *
 * Prefers the server-computed display_name; falls back to the canonical chain below.
 * See tests/fixtures/display-names.json for the authoritative table.
 */
export function getStarDisplayName(star: Star | StarDetail): string {
  // The server computes this too, and its value wins — see StarBase.display_name in
  // hygmap-api/app/schemas/star.py. This local chain is the fallback for rows that
  // arrive without one, and it must stay in step with the server and with PHP's
  // StarFormatter. All three are asserted against tests/fixtures/display-names.json;
  // change that table first.
  //
  // Canonical order (decided 2026-07-29): fictional > proper > bayer+con > flam+con >
  // GJ > HD > HIP > HR > CNS5 > TYC > Gaia > spect > ID. GJ leads because this is a map
  // of the solar neighbourhood; this chain used to lead with HIP.
  if (star.display_name) {
    return star.display_name
  }

  if (star.name) {
    return star.name
  }
  if (star.proper) {
    return star.proper
  }
  // Source data carries padding on bayer/flam, and the constellation is required —
  // a bare Greek letter is not a designation.
  if (star.bayer && star.con) {
    return `${star.bayer.trim()} ${star.con.trim()}`
  }
  if (star.flam && star.con) {
    return `${star.flam.trim()} ${star.con.trim()}`
  }

  const detail = star as StarDetail
  if (detail.gj) {
    return `GJ ${detail.gj}`
  }
  if (detail.hd) {
    return `HD ${detail.hd}`
  }
  if (detail.hip) {
    return `HIP ${detail.hip}`
  }
  if (detail.hr) {
    return `HR ${detail.hr}`
  }
  if (detail.cns5) {
    return `CNS5 ${detail.cns5}`
  }
  if (detail.tyc) {
    return `TYC ${detail.tyc}`
  }
  if (detail.gaia) {
    return `Gaia ${detail.gaia}`
  }
  if (star.spect) {
    return star.spect
  }

  return `ID ${star.id}`
}

